staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go

```diff
@@ -148,3 +169,277 @@ func setObjectMetaUsingAccessors(u, uCopy *unstructured.Unstructured) {
+	}
+
+	// Compare original unstructured vs final unstructured
+	tmp, ok = final.(runtime.Unstructured)
+	if !ok {
+		t.Fatalf("unexpected type %T for unstructured", tmp)
+	}
+	if !unstructuredEqual(t, original, final.(runtime.Unstructured)) {
+		t.Errorf("object changed during unstructured->A->unstructured->B->unstructured roundtrip, diff: %s", cmp.Diff(original, final))
+	}
+}
+
+func getSeed(t *testing.T) int64 {
+	seed := int64(time.Now().Nanosecond())
+	if override := os.Getenv("TEST_RAND_SEED"); len(override) > 0 {
+		overrideSeed, err := strconv.ParseInt(override, 10, 64)
+		if err != nil {
+			t.Fatal(err)
+		}
+		seed = overrideSeed
+		t.Logf("using overridden seed: %d", seed)
+	} else {
+		t.Logf("seed (override with TEST_RAND_SEED if desired): %d", seed)
+	}
+	return seed
+}
+
+const (
+	maxUnstructuredDepth  = 64
+	maxUnstructuredFanOut = 5
+)
+
+func unstructuredFuzzerFuncs(codecs serializer.CodecFactory) []interface{} {
+	return []interface{}{
+		func(u *unstructured.Unstructured, c fuzz.Continue) {
+			obj := make(map[string]interface{})
+			obj["apiVersion"] = generateValidAPIVersionString(c)
+			obj["kind"] = generateNonEmptyString(c)
+			for j := c.Intn(maxUnstructuredFanOut); j >= 0; j-- {
+				obj[c.RandString()] = generateRandomTypeValue(maxUnstructuredDepth, c)
+			}
+			u.Object = obj
+		},
+		func(ul *unstructured.UnstructuredList, c fuzz.Continue) {
+			obj := make(map[string]interface{})
+			obj["apiVersion"] = generateValidAPIVersionString(c)
+			obj["kind"] = generateNonEmptyString(c)
+			for j := c.Intn(maxUnstructuredFanOut); j >= 0; j-- {
+				obj[c.RandString()] = generateRandomTypeValue(maxUnstructuredDepth, c)
+			}
+			for j := c.Intn(maxUnstructuredFanOut); j >= 0; j-- {
+				var item = unstructured.Unstructured{}
+				c.Fuzz(&item)
+				ul.Items = append(ul.Items, item)
+			}
+			ul.Object = obj
+		},
+	}
+}
+
+func generateNonEmptyString(c fuzz.Continue) string {
+	temp := c.RandString()
+	for len(temp) == 0 {
+		temp = c.RandString()
+	}
+	return temp
+}
+
+// generateNonEmptyNoSlashString generates a non-empty string without any slashes
+func generateNonEmptyNoSlashString(c fuzz.Continue) string {
+	temp := strings.ReplaceAll(generateNonEmptyString(c), "/", "")
+	for len(temp) == 0 {
+		temp = strings.ReplaceAll(generateNonEmptyString(c), "/", "")
+	}
+	return temp
+}
+
+// generateValidAPIVersionString generates valid apiVersion string with formats:
+// <string>/<string> or <string>
+func generateValidAPIVersionString(c fuzz.Continue) string {
+	if c.RandBool() {
+		return generateNonEmptyNoSlashString(c) + "/" + generateNonEmptyNoSlashString(c)
+	} else {
+		return generateNonEmptyNoSlashString(c)
+	}
+}
+
+func generateRandomTypeValue(depth int, c fuzz.Continue) interface{} {
+	t := c.Rand.Intn(120)
+	// If the max depth for unstructured is reached, only add non-recursive types
+	// which is 20+ in range
+	if depth == 0 {
+		t = 20 + c.Rand.Intn(120-20)
+	}
+
+	switch {
```

Comment:
metav1.Time stands out as a missing primitive.  Probably a bigInt and bigDouble as well.  Also a duration, an intOrString, and a resource.Quantity.  Another standard specials I'm missing?

Reply:
We only generate/fuzz valid JSON data types such as map[string]interface{}, []interface{}, int64, float64, string, boolean and nil (null) because when decoding into unstructured can only produce those supported types.

If a valid from other types into an unstructured, it will roundtrip to one of the above list of supported types. For example, if Time type is used, it will be encoded into a ISO 8601 format string such as "2001-02-03T12:34:56Z" which as far as fuzz testing is concerned, it is just a string.

Also, there is another fuzz test ([RoundtripToUnstructured](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/unstructured.go#L51)) for Unstructured objects that covers some additional k8s types including metav1.Time.

I added comments for this `generateRandomTypeValue` func to capture the info that mentioned in this comment. I hope that is satisfiable for you.